libkatana_python_native/include/katana/python/NumbaSupport.h

```diff
@@ -0,0 +1,146 @@
+/// calls to the function via a static function pointer allowing it to be
+/// called from Numba generated code.
+///
+/// @tparam Cls The class containing the method.
+/// @tparam Return The return value of the method.
+/// @tparam Args The argument types to the method.
+template <typename Cls, typename Return, typename... Args>
+struct MemberFunction {
+  Return (Cls::*member_func)(Args...);
+
+  static remove_cvref_t<Return> call(
+      MemberFunction<Cls, Return, Args...>* func, Cls* self,
+      remove_cvref_t<Args>... args) {
+    return (self->*(func->member_func))(args...);
+  }
+};
+
+template <typename Cls, typename Return, typename... Args>
+struct ConstMemberFunction {
+  Return (Cls::*member_func)(Args...) const;
+
+  static remove_cvref_t<Return> call(
+      ConstMemberFunction<Cls, Return, Args...>* func, const Cls* self,
+      remove_cvref_t<Args>... args) {
+    return (self->*(func->member_func))(args...);
+  }
+};
+
+template <typename... Args>
+class DefWithNumbaImpl {
+  template <
+      typename Cls, typename Return, typename Func, typename Caller,
+      typename... Extra>
+  void def_class_method(
+      pybind11::class_<Cls>& cls, const char* name, Func f, Caller* caller,
+      const Extra&... extra) const {
+    cls.def(name, f, extra...);
+    auto numba_support =
+        pybind11::module_::import("katana.native_interfacing.numba_support");
+    numba_support.attr("register_method")(
+        cls, cls.attr(name), (uintptr_t)&Caller::call, (uintptr_t)caller,
+        PythonTypeTraits<remove_cvref_t<Return>>::ctypes_type(),
+        PythonTypeTraits<remove_cvref_t<Args>>::ctypes_type()...);
+  }
+
+  template <typename Return, typename... Extra>
+  void def_func(
+      pybind11::module_& m, const char* name, Return (*f)(Args...),
+      const Extra&... extra) const {
+    m.def(name, f, extra...);
+    auto func = m.attr(name);
+    auto numba_support =
+        pybind11::module_::import("katana.native_interfacing.numba_support");
+    numba_support.attr("register_function")(
+        func, (uintptr_t)f, 0,
+        PythonTypeTraits<remove_cvref_t<Return>>::ctypes_type(),
+        PythonTypeTraits<remove_cvref_t<Args>>::ctypes_type()...);
+  }
+
+public:
+  // TODO(amp): The generated wrappers used from numba code are created per
+  //  *signature*, not per function. So two functions will the same overall
+  //  signature (including containing class) will use the same wrapper. This
+  //  could possibly cause a megamorphic call site that defeats the CPU branch
+  //  predictor. The fix for this is to somehow distinguish different
+  //  functions/methods at the template level. This could be done with a counter
+  //  or a name argument of some kind.
+
+  template <typename Return, typename... Extra>
+  void operator()(
+      pybind11::module_& m, const char* name, Return (*pf)(Args...),
+      const Extra&... extra) const {
+    def_func(m, name, pf, extra...);
+  }
```

Comment:
When is this function called?

Reply:
This particular overload may not be used yet. In fact I'm pretty sure it isn't. What is allows for is:

```
DefWithNumba(cls, "method", [](Cls* self) { return self->thing() + 1; });
```

This allows lambdas or static methods that make small changes to the C++ method semantics.